components/brave_rewards/core/ledger_impl.h

```diff
@@ -303,7 +303,7 @@ class LedgerImpl : public mojom::Ledger {
 
   template <typename T>
   T GetState(const std::string& name) {
-    T value;
+    T value{};
```

Comment:
@szilardszaloki My theory is that this previously uninitialized variable (along with odd sync mojo behavior) is why we are seeing the debug crashes in `GetRewardsParameters`.

Reply:
@zenparsing Exactly one of the paths below should always initialize `value`. Can surely default initialize it, but unless the corresponding sync call fails, this variable should be initialized before `GetState()` returns.

Reply:
Default initializing `value` rather masks the problem, since it would deceptively return a default value when a sync Mojom call fails, and we'll never find out about it. The right approach would be to use `absl::optional<T>`, and return `absl::nullopt` on Mojom failures, and the value otherwise.

Reply:
Here's the Mojom code:
```cpp
bool LedgerClientProxy::GetDoubleState(
    const std::string& param_name, double* out_param_value) {
#if BUILDFLAG(MOJO_TRACE_ENABLED)
  TRACE_EVENT_BEGIN1(
    "mojom", "Call brave_rewards::mojom::LedgerClient::GetDoubleState (sync)", "input_parameters",
    [&](perfetto::TracedValue context){
      auto dict = std::move(context).WriteDictionary();
      perfetto::WriteIntoTracedValueWithFallback(
           dict.AddItem("name"), param_name,
                        "<value of type const std::string&>");
   });
#else
  TRACE_EVENT0("mojom", "LedgerClient::GetDoubleState");
#endif
  const bool kExpectsResponse = true;
  const bool kIsSync = true;
  const bool kAllowInterrupt =
      true;
  
  const uint32_t kFlags =
      ((kExpectsResponse) ? mojo::Message::kFlagExpectsResponse : 0) |
      ((kIsSync) ? mojo::Message::kFlagIsSync : 0) |
      ((kAllowInterrupt) ? 0 : mojo::Message::kFlagNoInterrupt);
  
  mojo::Message message(
      internal::kLedgerClient_GetDoubleState_Name, kFlags, 0, 0, nullptr);
  mojo::internal::MessageFragment<
      ::brave_rewards::mojom::internal::LedgerClient_GetDoubleState_Params_Data> params(
          message);
  params.Allocate();
  mojo::internal::MessageFragment<
      typename decltype(params->name)::BaseType> name_fragment(
          params.message());
  mojo::internal::Serialize<mojo::StringDataView>(
      param_name, name_fragment);
  params->name.Set(
      name_fragment.is_null() ? nullptr : name_fragment.data());
  MOJO_INTERNAL_DLOG_SERIALIZATION_WARNING(
      params->name.is_null(),
      mojo::internal::VALIDATION_ERROR_UNEXPECTED_NULL_POINTER,
      "null name in LedgerClient.GetDoubleState request");

#if defined(ENABLE_IPC_FUZZER)
  message.set_interface_name(LedgerClient::Name_);
  message.set_method_name("GetDoubleState");
#endif

  bool result = false;
  std::unique_ptr<mojo::MessageReceiver> responder(
      new LedgerClient_GetDoubleState_HandleSyncResponse(
          &result, out_param_value));
  ::mojo::internal::SendMojoMessage(*receiver_, message, std::move(responder));
#if BUILDFLAG(MOJO_TRACE_ENABLED)
  TRACE_EVENT_END1(
    "mojom", "LedgerClient::GetDoubleState", "sync_response_parameters",
    [&](perfetto::TracedValue context){
      auto dict = std::move(context).WriteDictionary();
      perfetto::WriteIntoTracedValueWithFallback(
           dict.AddItem("value"), out_param_value,
                        "<value of type double>");
   });
#endif
  return result;
}
```
Relevant part is the last 17 lines.
Also:
```cpp
class LedgerClient_GetDoubleState_HandleSyncResponse
    : public mojo::MessageReceiver {
 public:
  LedgerClient_GetDoubleState_HandleSyncResponse(
      bool* result, double* out_value)
      : result_(result), out_value_(out_value) {
    DCHECK(!*result_);
  }

  LedgerClient_GetDoubleState_HandleSyncResponse(const LedgerClient_GetDoubleState_HandleSyncResponse&) = delete;
  LedgerClient_GetDoubleState_HandleSyncResponse& operator=(const LedgerClient_GetDoubleState_HandleSyncResponse&) = delete;

  bool Accept(mojo::Message* message) override;
 private:
  bool* result_;
  double* out_value_;};

bool LedgerClient_GetDoubleState_HandleSyncResponse::Accept(
    mojo::Message* message) {

  DCHECK(message->is_serialized());
  internal::LedgerClient_GetDoubleState_ResponseParams_Data* params =
      reinterpret_cast<internal::LedgerClient_GetDoubleState_ResponseParams_Data*>(
          message->mutable_payload());
  
  bool success = true;
  double p_value{};
  LedgerClient_GetDoubleState_ResponseParamsDataView input_data_view(params, message);
  
  if (success)
    p_value = input_data_view.value();
  if (!success) {
    ReportValidationErrorForMessage(
        message,
        mojo::internal::VALIDATION_ERROR_DESERIALIZATION_FAILED,
        LedgerClient::Name_, 13, true);
    return false;
  }
  *out_value_ = std::move(p_value);
  *result_ = true;
  return true;
}
```
From the above it seems that the generated code for `LedgerClient_GetDoubleState_HandleSyncResponse::Accept()` will always return `true` (`success` doesn't depend on any validation logic, i.e. no built-in Mojom validation for `double`s), hence `*result_` is always set to `true`. This means that `LedgerClientProxy::GetDoubleState()` will also return `true`:
```cpp
  bool result = false;
  std::unique_ptr<mojo::MessageReceiver> responder(
      new LedgerClient_GetDoubleState_HandleSyncResponse(
          &result, out_param_value));
  ::mojo::internal::SendMojoMessage(*receiver_, message, std::move(responder));
#if BUILDFLAG(MOJO_TRACE_ENABLED)
  TRACE_EVENT_END1(
    "mojom", "LedgerClient::GetDoubleState", "sync_response_parameters",
    [&](perfetto::TracedValue context){
      auto dict = std::move(context).WriteDictionary();
      perfetto::WriteIntoTracedValueWithFallback(
           dict.AddItem("value"), out_param_value,
                        "<value of type double>");
   });
#endif
  return result;
```
therefore the initialization of `value` in our code always has to take place.

Reply:
What happens if the other end is no longer there (the client has disconnected)?

Reply:
You mean, if we call `reset()` on the `mojo::AssociatedReceiver<mojom::LedgerClient>` in `RewardsServiceImpl`? The `mojo::AssociatedRemote<mojom::LedgerClient>` in the ledger doesn't have a disconnect handler, so I guess this boils down to the question whether you get back `true`, or `false` when calling a sync function on a bound but disconnected `mojo::Remote<>` — you get back `false`, and the variable is left uninitialized.

Currently, this can happen as an edge case of `RewardsServiceImpl::Shutdown()`, but again, our Mojom setup needs fixes anyway, as this is not the only problem. Let's reduce the `Get*State`/`Set*State` APIs to a single pair with `base::Value`, which return an `absl::optional<>` to indicate Mojom sync call failures. Then both problems go away (well, technically the "uninitialized variable" part is not the problem here, as it's a commonly used pattern where a default value does not make sense, and where you initialize your variable on all the possible paths).

Reply:
> well, technically the "uninitialized variable" part is not the problem here, as it's a commonly used pattern where a default value does not make sense, and where you initialize your variable on all the possible paths

It's far better to get a known "wrong" default than a random value (which is simply baffling).

Reply:
Yeah, that's unquestionable, in general. I'm just saying that we wouldn't have had this conversation ever, if we'd used a default value here, and that initializing `value` via `{}` does not fix the code even now, just gets rid of the symptom, as the "no/invalid value" case cannot be represented by the default value (which is perfectly valid and hides the underlying problem).

Reply:
As an alternative (if you find the `absl::optional<>` version painful to use — as it would indeed add a lot of branching):
```cpp
template <typename T>
T GetState(const std::string& name) {
  T value;
  bool success = false;

  if constexpr (std::is_same_v<T, bool>) {
    success = ledger_client_->GetBooleanState(name, &value);
  } else if constexpr (std::is_same_v<T, int32_t>) {
    success = ledger_client_->GetIntegerState(name, &value);
  } else if constexpr (std::is_same_v<T, double>) {
    success = ledger_client_->GetDoubleState(name, &value);
  } else if constexpr (std::is_same_v<T, std::string>) {
    success = ledger_client_->GetStringState(name, &value);
  } else if constexpr (std::is_same_v<T, int64_t>) {
    success = ledger_client_->GetInt64State(name, &value);
  } else if constexpr (std::is_same_v<T, uint64_t>) {
    success = ledger_client_->GetUint64State(name, &value);
  } else if constexpr (std::is_same_v<T, base::Value>) {
    success = ledger_client_->GetValueState(name, &value);
  } else if constexpr (std::is_same_v<T, base::Time>) {
    success = ledger_client_->GetTimeState(name, &value);
  } else {
    static_assert(base::AlwaysFalse<T>, "Unsupported type!");
  }

  CHECK(success) << "Something went wrong while reading the pref " << name;
  return value;
}
```
and fix our underlying Mojom issues.

Reply:
The `CHECK` is a good idea - I was able to confirm that these sync calls can fail during shutdown, and therefore return uninitialized values. This I believe explains the failures in `GetRewardsParameters`.